src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/PairReadsAndVariants.java

```diff
@@ -0,0 +1,87 @@
+package org.broadinstitute.hellbender.engine.dataflow.transforms;
+
+import com.google.cloud.dataflow.sdk.transforms.*;
+import com.google.cloud.dataflow.sdk.transforms.join.CoGbkResult;
+import com.google.cloud.dataflow.sdk.transforms.join.CoGroupByKey;
+import com.google.cloud.dataflow.sdk.transforms.join.KeyedPCollectionTuple;
+import com.google.cloud.dataflow.sdk.values.KV;
+import com.google.cloud.dataflow.sdk.values.PCollection;
+import com.google.cloud.dataflow.sdk.values.TupleTag;
+import org.broadinstitute.hellbender.engine.dataflow.datasources.VariantShard;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.variant.Variant;
+
+/**
+ * PairReadsAndVariants takes two PCollections (GATKRead and Variant) and create a single PCollection of KV where
+ * GATKRead is the key and Variant is the value. We do this join by sharding both collections by "variant shard" and
+ * checking for overlap on each "shard."
+ * This currently will not work with unmapped reads.
+ *
+ * step 1: key reads and variants by shard
+ * |---- shard 0 -----|---- shard 1 -----|---- shard 2 -----|---- shard 3 -----|---- shard 4 -----|
+ *     |---------- read a ---------|               |----- read b ------|
+ *   |- variant 1 -|    |- variant 2 -|               |- variant 3 -|
+ *
+ * step 2: shard read and variant by variant shard
+ *                      |---- shard 0 -----|
+ *                          |---------- read a ---------|
+ *                        |- variant 1 -|
+ *
+ *
+ *                      |---- shard 1 -----|
+ *       |---------- read a ---------|
+ *                        |- variant 2 -|
+ *
+ *
+ *                      |---- shard 2 -----|
+ *                                |----- read b ------|
+ *                                   |- variant 3 -|
+ *
+ *                      |---- shard 3 -----|
+ *             |----- read b ------|
+ *                |- variant 3 -|
+ *
+ * step 3: pair reads and variants
+ * KV<read a, variant 1> // from shard 0
+ * KV<read a, variant 2> // from shard 1
+ * KV<read b, variant 3> // from shard 2
+ * KV<read b, variant 3> // from shard 3
+ *
+ */
+public class PairReadsAndVariants {
+    static public PCollection<KV<GATKRead, Variant>> pair(PCollection<GATKRead> pRead, PCollection<Variant> pVariant) {
```

Comment:
Would it be useful to use generics instead, as this class actually applies to any Interval-like class, not just reads and variants? This way we could get overlapping pairs of reference shards as well if we wanted, or of (Variant,GATKRead) if we wanted to key by variant later.